java/ql/src/Security/CWE/CWE-470/ReflectionBad.ql

```diff
@@ -0,0 +1,54 @@
+/**
+ * @name Unsafe reflection
+ * @description External input with reflection to select which classes to use
+ * @kind reflection
+ * @problem.severity warning
+ * @precision medium
+ * @id java/unsafe-reflection
+ * @tags security
+ *       external/cwe/cwe-470
+ */
+import java
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+class ClassFornameMethod extends Method {
+  ClassFornameMethod() {
+    this.getDeclaringType().hasQualifiedName("java.lang", "Class<>") and
```

Comment:
Is this right? Shouldn't this be this instead?

```suggestion
    this.getDeclaringType().hasQualifiedName("java.lang", "Class") and
```

If not, what does the `Class<>` actually do?

Reply:
Class in java is defined as Class<T>. And to the qualified name for that in codeql is `Class<>` @JLLeitschuh

Reply:
Actually, the method with `.getDeclaringType().hasQualifiedName("java.lang", "Class")` is the source declaration, whereas the type with `.hasQualifiedName("java.lang", "Class<>")` is the raw type.  So when we want to match all the methods on all the instantiations of `Class<T>`, then we need to do `this.getDeclaringType().getSourceDeclaration().hasQualifiedName("java.lang", "Class")`

Reply:
Given this confusion, I think a test would be a good to verify this logic actually works.